### PR TITLE
Port audit_desc audit rules to a cop

### PR DESCRIPTION
**Pull request by GauthamGoli**
- [x] Have you followed the guidelines in our [Contributing](https://github.com/Homebrew/brew/blob/master/CONTRIBUTING.md) document?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/Homebrew/brew/pulls) for the same change?
- [x] Have you added an explanation of what your changes do and why you'd like us to include them?
- [x] Have you written new tests for your changes? [Here's an example](https://github.com/Homebrew/homebrew/pull/49031).
- [x] Have you successfully run `brew tests` with your changes locally?

-----
This is in continuation of the PR #1873 and issue #569 
This PR ports `audit_desc` audit rules into a rubocop custom cop. The rules are 

- Check if `desc` exists in a Formula
- Check if `desc` size is > 80
- Check if `desc` begins with an article
- Check for proper usage of `command-line` in `desc`
- Check if `desc` contains the formula name

The previous tests for the ported audit rules in `audit_spec.rb` have been removed. ~~I am working on adding tests for the Custom Cops. I have the basic code, but I'm currently trying to figure out where to place the code for the tests to get executed.~~
New tests for the RuboCop custom cops have been added and `Gemfile` has been updated.

### CODE REVIEW

**Comment by GauthamGoli** [comment]
`brew style` without any args runs style checks on whole Homebrew `Library` which is why the build is failing as custom cops are running on files outside `Library/Taps/homebrew/`.

Instead we want Custom Cops to be run only on formulae files. Can I modify `style.rb` to achieve this? (To exclude custom cop checks on core code and include them on formulae)

**Comment by reitermarkus** [comment]
`rubocop-cask` is checking checking if the file is a Cask,

https://github.com/caskroom/rubocop-cask/blob/eca041681dc554996a926173ec1d0c378efaecd3/lib/rubocop/cask/extend/node.rb#L14-L20


and skips it if it isn't.

https://github.com/caskroom/rubocop-cask/blob/cd020379b3a38d4dc20a23eb6c596b5f39414565/lib/rubocop/cop/cask/mixin/cask_help.rb#L9

**Comment by GauthamGoli** [comment]
@reitermarkus I too added a check for the same https://github.com/Homebrew/brew/pull/2242/commits/19f693d25ba514a6b604128474763e83de55b548#diff-166fafc45d1d65494760d2fb578d6e46R8

But the problem is there are formulae defined outside `Library/Taps/homebrew/` https://travis-ci.org/Homebrew/brew/jobs/206899446#L266-L279

**Comment by reitermarkus** [comment]
> I too added a check for the same

Ok, so there is a `on_class` method, do you know if there is a `on_file` method? This way we could simply skip all files under `Library/Homebrew`.

**Comment by reitermarkus** [comment]
This looks like it could do what we want: https://github.com/bbatsov/rubocop/blob/9754e06f468631e9145b72f30e23e4cc0883bdb8/lib/rubocop/cop/style/file_name.rb#L20-L27

**Comment by GauthamGoli** [comment]
@reitermarkus Yes, just found that too. Thanks! :)

File: Library/Homebrew/rubocops/formula_desc_cop.rb [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+  module Cop
+    module Homebrew
+      class FormulaDesc < Cop
+        def on_class(node)
+          return unless processed_source.path =~ %r{#{ENV["HOMEBREW_LIBRARY"]}\/Taps\/homebrew}
+          class_node, parent_class_node, body = *node
+          formula_name = class_node.const_name
+          return unless parent_class_node && parent_class_node.const_name == "Formula" && body
+          check(node, body, formula_name)
+        end
```
Comment (reitermarkus) [inline Library/Homebrew/rubocops/formula_desc_cop.rb]: Can you move this to a helper method? We'll probably need this for every Formula cop.

File: Library/Homebrew/rubocops/mixins/on_formula_class.rb [hunk 2]
```diff
@@ -0,0 +1,18 @@
+module RuboCop
+  module Cop
+    # Helper module for checking formula class
+    module FormulaDef
```
Comment (MikeMcQuaid) [inline Library/Homebrew/rubocops/mixins/on_formula_class.rb]: To me it makes more sense for there to be a `HomebrewCop` base class that other cops inherit from as all will want to use this code.

File: Library/Homebrew/rubocops/mixins/on_formula_class.rb [hunk 3]
```diff
@@ -0,0 +1,18 @@
...
+      def file_inside_formula_root?
+        return false unless processed_source.path =~ %r{#{ENV["HOMEBREW_LIBRARY"]}\/Taps\/homebrew}
```
Comment (MikeMcQuaid) [inline Library/Homebrew/rubocops/mixins/on_formula_class.rb]: Don't think `HOMEBREW_LIBRARY` will be set when using `rubocop` outside of `brew style`?

File: Library/Homebrew/rubocops/formula_desc_cop.rb [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+            line_begin_pos = call_node.source_range.source_buffer.line_range(call_node.loc.line).begin_pos
+            desc_begin_pos = call_node.children[2].source_range.begin_pos
+
+            linelength = "#{formula_name}: #{description}".length
+            if linelength > 80
+              column = desc_begin_pos - line_begin_pos
+              length = call_node.children[2].source_range.size
+              sourcerange = source_range(source_buffer, line_number, column, length)
+              message = <<-EOS.strip_indent
+                        Description is too long. \"name: desc\" should be less than 80 characters.
```
Comment (MikeMcQuaid) [inline Library/Homebrew/rubocops/formula_desc_cop.rb]: Indent this two characters after `message` and use `EOS.undent`. Are you sure you need to escape `\"`?

File: Library/Homebrew/rubocops/formula_desc_cop.rb [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+        def on_formula(node, class_node, _parent_class_node, body)
+          check(node, body, class_node.const_name)
+        end
+
+        private
+
+        def check(node, body, formula_name)
+          body.each_child_node(:send) do |call_node|
+            _receiver, call_name, args = *call_node
+            next unless call_name == :desc && !args.children[0].empty?
```
Comment (MikeMcQuaid) [inline Library/Homebrew/rubocops/formula_desc_cop.rb]: `unless &&` is generally hard to parse; can you flip this to an `if`?

File: Library/Homebrew/rubocops/formula_desc_cop.rb [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+              length = call_node.children[2].source_range.size
+              sourcerange = source_range(source_buffer, line_number, column, length)
+              message = <<-EOS.strip_indent
+                        Description is too long. \"name: desc\" should be less than 80 characters.
+                        Length is calculated as #{formula_name} + desc. (currently #{linelength})
+                        EOS
+              add_offense(call_node, sourcerange, message)
+            end
+
+            match_object = description.match(/([Cc]ommand ?line)/)
```
Comment (MikeMcQuaid) [inline Library/Homebrew/rubocops/formula_desc_cop.rb]: Could make this regex case-insensitive?

File: Library/Homebrew/rubocops/mixins/on_formula_class.rb (hunk 2 above)
Comment (GauthamGoli) [inline Library/Homebrew/rubocops/mixins/on_formula_class.rb]: Yeah, this makes more sense, but I was not able to inherit `HomebrewCop` due to some error. Found a fix. Will push the changes.

File: Library/Homebrew/rubocops/mixins/on_formula_class.rb (hunk 3 above)
Comment (GauthamGoli) [inline Library/Homebrew/rubocops/mixins/on_formula_class.rb]: Should I use `%r{\/Taps\/homebrew}` without `HOMEBREW_LIBRARY` ?

File: Library/Homebrew/rubocops/extend/Homebrewcop.rb [hunk 7]
```diff
@@ -0,0 +1,19 @@
+module RuboCop
+  module Cop
+    class HomebrewCop < Cop
```
Comment (reitermarkus) [inline Library/Homebrew/rubocops/extend/Homebrewcop.rb]: I'd call this `FormulaCop`. Also, use `snake_case` for the filename.

Comment (GauthamGoli) [inline Library/Homebrew/rubocops/extend/Homebrewcop.rb]: Thanks.

File: Library/Homebrew/rubocops/mixins/on_formula_class.rb (hunk 3 above)
Comment (MikeMcQuaid) [inline Library/Homebrew/rubocops/mixins/on_formula_class.rb]: Yeh, that sounds like a sensible default. Don't need `\/` inside a `%r`, though!

File: Library/Homebrew/rubocops/bottle_block_cop.rb [hunk 8]
```diff
@@ -1,20 +1,31 @@
...
 
-          return unless method_name == :bottle
-          check_revision?(body)
+        def on_formula(_node, _class_node, _parent_class_node, body)
+          check(body)
         end
 
         private
 
+        def check(body)
```
Comment (MikeMcQuaid) [inline Library/Homebrew/rubocops/bottle_block_cop.rb]: Not totally clear to me what `body` is here; could you explain so we can maybe pick a better name?

File: Library/Homebrew/rubocops/bottle_block_cop.rb [hunk 9]
```diff
@@ -1,20 +1,31 @@
...
+          body.each_child_node(:block) do |block_node|
```
Comment (MikeMcQuaid) [inline Library/Homebrew/rubocops/bottle_block_cop.rb]: Not totally clear to me what `block_node ` is here; could you explain so we can maybe pick a better name?

File: Library/Homebrew/rubocops/extend/formula_cop.rb [hunk 10]
```diff
@@ -0,0 +1,19 @@
+module RuboCop
+  module Cop
+    class FormulaCop < Cop
+      @registry = Cop.registry
+
+      def file_inside_formula_root?
+        return false unless processed_source.path =~ %r{/Taps/homebrew}
+        true
```
Comment (MikeMcQuaid) [inline Library/Homebrew/rubocops/extend/formula_cop.rb]: `!(processed_source.path =~ %r{/Taps/homebrew}).nil?` maybe?

File: Library/Homebrew/rubocops/extend/formula_cop.rb [hunk 11]
```diff
@@ -0,0 +1,19 @@
...
+      end
+
+      def on_class(node)
```
Comment (MikeMcQuaid) [inline Library/Homebrew/rubocops/extend/formula_cop.rb]: Not sure what `node` is here; could you explain so we can figure out a better name?

File: Library/Homebrew/rubocops/extend/formula_cop.rb [hunk 12]
```diff
@@ -0,0 +1,19 @@
...
+        return unless file_inside_formula_root?
```
Comment (MikeMcQuaid) [inline Library/Homebrew/rubocops/extend/formula_cop.rb]: I wonder if this could be done in a shared initializer?

File: Library/Homebrew/rubocops/formula_desc_cop.rb [hunk 13]
```diff
@@ -0,0 +1,75 @@
...
+      # This cop audits `desc` in Formulae
+      #
+      # - Checks for existence of `desc`
+      # - Checks if size of `desc` > 80
+      # - Checks if `desc` begins with an article
+      # - Checks for correct usage of `command-line` in `desc`
+      # - Checks if `desc` contains the formula name
+
+      class FormulaDesc < FormulaCop
+        def on_formula(node, class_node, _parent_class_node, body)
```
Comment (MikeMcQuaid) [inline Library/Homebrew/rubocops/formula_desc_cop.rb]: Not super clear what `on_formula` means or the various `node`s?

File: Library/Homebrew/rubocops/formula_desc_cop.rb [hunk 14]
```diff
@@ -0,0 +1,75 @@
...
+          check(node, body, class_node.const_name)
+        end
+
+        private
+
+        def check(node, body, formula_name)
```
Comment (MikeMcQuaid) [inline Library/Homebrew/rubocops/formula_desc_cop.rb]: Would be good to figure out a way to extract more of the Rubocop/Ruby parse tree logic here from the audit logic.

File: Library/Homebrew/rubocops/formula_desc_cop.rb [hunk 15]
```diff
@@ -0,0 +1,75 @@
...
+              EOS
+              add_offense(call_node, sourcerange, message)
+            end
+
+            match_object = description.match(/(command ?line)/i)
+            if match_object
+              column = desc_begin_pos+match_object.begin(0)-line_begin_pos+1
+              length = match_object.to_s.length
+              sourcerange = source_range(source_buffer, line_number, column, length)
+              add_offense(call_node, sourcerange, "Description should use \"command-line\" instead of \"#{match_object}\"")
```
Comment (MikeMcQuaid) [inline Library/Homebrew/rubocops/formula_desc_cop.rb]: A linebreak here would be good to make this a little shorter

File: Library/Homebrew/rubocops/formula_desc_cop.rb (hunk 13 above)
Comment (GauthamGoli) [inline Library/Homebrew/rubocops/formula_desc_cop.rb]: Inside `FormulaCop` there's a method `on_class` which does the following sequentially

-  Checks if the file being processed is inside /Taps/Homebrew/
-  If so, check if current class is being inherited from `Formula` ( Thus making sure that its a Homebrew Formula)
-  Call `on_formula` with arguments below. (`on_formula` is not yet defined as we will define it inside the CustomCop which inherits from `FormulaCop`)
-  `node` is the RuboCop::AST::node instance of whole formula
-  `class_node` is the RuboCop::AST::node instance of formula's name
-  `_parent_class_node` is the RuboCop::AST::node instance of `Formula` 
-  `body` is the RubCop::AST::node for ever thing thats inside the Formula

`class_node`, `_parent_class_node`, `body` are children of `node` and may be we can omit passing `node` as the argument to `on_formula` as its redundant.

File: Library/Homebrew/rubocops/extend/formula_cop.rb (hunk 11 above)
Comment (GauthamGoli) [inline Library/Homebrew/rubocops/extend/formula_cop.rb]: RuboCop processes the source code and creates an Abstract Syntax Tree(AST) representation, and traverses the AST. As it encounters a `node` of some `type`, a message is sent to cop instance class hierarchy `send(:"on_#{node.type}", node)`
and we override it by defining `on_class` to check and add violations.

File: Library/Homebrew/rubocops/extend/formula_cop.rb (hunk 10 above)
Comment (GauthamGoli) [inline Library/Homebrew/rubocops/extend/formula_cop.rb]: Yes, This is more elegant 👍

File: Library/Homebrew/rubocops/bottle_block_cop.rb (hunk 9 above)
Comment (GauthamGoli) [inline Library/Homebrew/rubocops/bottle_block_cop.rb]: `body.each_child_node(:block)` calls the given block for each `child` node which has `child.type==:block`. (We want to iterate only over blocks and find `bottle` block). 
The given block for `body.each_child_node(:block)` takes `block_node` as the argument.

File: Library/Homebrew/rubocops/bottle_block_cop.rb (hunk 8 above)
Comment (GauthamGoli) [inline Library/Homebrew/rubocops/bottle_block_cop.rb]: `body` is the `RuboCop::AST::Node` instance of everything that's inside the formula's class.
The child nodes of `body` are also instances of `RuboCop::AST::Node`

File: Library/Homebrew/rubocops/extend/formula_cop.rb (hunk 12 above)
Comment (GauthamGoli) [inline Library/Homebrew/rubocops/extend/formula_cop.rb]: Can you please explain what's a shared initializer and why would we want to use it?

File: Library/Homebrew/rubocops/formula_desc_cop.rb (hunk 13 above)
Comment (MikeMcQuaid) [inline Library/Homebrew/rubocops/formula_desc_cop.rb]: I'm wondering if it's worth being called `audit_formula`; the `on_` naming scheme is better if people are familiar with RuboCop but I'd think the majority won't be.

File: Library/Homebrew/rubocops/formula_desc_cop.rb (hunk 14 above)
Comment (MikeMcQuaid) [inline Library/Homebrew/rubocops/formula_desc_cop.rb]: Perhaps by generally splitting this into a bunch of smaller methods.

File: Library/Homebrew/rubocops/extend/formula_cop.rb (hunk 10 above)
Comment (reitermarkus) [inline Library/Homebrew/rubocops/extend/formula_cop.rb]: This will exclude Formulae in all Taps other than `homebrew/*`.

Comment (GauthamGoli) [inline Library/Homebrew/rubocops/extend/formula_cop.rb]: Actually, this may not be necessary. We can use `Include` and `Exclude` options for individual cops in `.rubocops.yml` to pass file/directory patterns to decide whether cops will run or not. I will update the PR. 
@reitermarkus Where else are the Formulae located?

Comment (reitermarkus) [inline Library/Homebrew/rubocops/extend/formula_cop.rb]: > Where else are the Formulae located?

In personal/custom Taps, and I think you can also pass Ruby files directly to `brew style`.

Comment (GauthamGoli) [inline Library/Homebrew/rubocops/extend/formula_cop.rb]: Yeah you're right. Thanks. I'll think of some way to handle this.

File: Library/Homebrew/rubocops/bottle_block_cop.rb [hunk 16]
```diff
@@ -1,20 +1,31 @@
...
+            next if block_length(block_node).zero?
+            method, _args, block_body = *block_node
```
Comment (GauthamGoli) [inline Library/Homebrew/rubocops/bottle_block_cop.rb]: If there are multiple such unused variables in the same block, instead of prefixing `_` for all of them, using just `_` , would it be okay?

Comment (MikeMcQuaid) [inline Library/Homebrew/rubocops/bottle_block_cop.rb]: Yeh, I think using `_` instead of all of them would be good 👍

File: Library/Homebrew/rubocops/formula_desc_cop.rb (hunk 14 above)
Comment (GauthamGoli) [inline Library/Homebrew/rubocops/formula_desc_cop.rb]: While trying to split them into smaller methods, `line_number`, `line_begin_pos`, `desc_begin_pos`, `call_node` are required as arguments in every method. Can I make them instance variables?

Comment (reitermarkus) [inline Library/Homebrew/rubocops/formula_desc_cop.rb]: Can you push the changes? Otherwise it's hard to tell if instance variables make sense.

Comment (MikeMcQuaid) [inline Library/Homebrew/rubocops/formula_desc_cop.rb]: Agreed; go for it.

Comment (GauthamGoli) [inline Library/Homebrew/rubocops/formula_desc_cop.rb]: I've made an additional commit and also some style checks won't be passing. I will squash the commits later once code gets reviewed.

File: Library/Homebrew/rubocops/formula_desc_cop.rb [hunk 17]
```diff
@@ -0,0 +1,84 @@
...
+                      :desc_begin_pos, :call_node
+
+        def audit_formula(node, class_node, _parent_class_node, body)
+          check(node, body, class_node.const_name)
+        end
+
+        private
+
+        def check(node, body, formula_name)
+          body.each_child_node(:send) do |call_node|
```
Comment (reitermarkus) [inline Library/Homebrew/rubocops/formula_desc_cop.rb]: You can make this `do |(_receiver, call_name, args)|`.

File: Library/.rubocop.yml [hunk 18]
```diff
@@ -8,7 +8,17 @@ AllCops:
 
 require: ./Homebrew/rubocops.rb
 
+Homebrew/FormulaCop:
```
Comment (MikeMcQuaid) [inline Library/.rubocop.yml]: Maybe worth putting this in another directory so it doesn't need to be disabled like this?

File: Library/.rubocop.yml [hunk 19]
```diff
@@ -8,7 +8,17 @@ AllCops:
...
+  Enabled: false
+
 Homebrew/CorrectBottleBlock:
+  Include:
+    - '**/Taps/homebrew/**/*.rb'
```
Comment (MikeMcQuaid) [inline Library/.rubocop.yml]: If this will need to be repeated for every cop I think the existing approach was superior.

File: Library/Homebrew/rubocops/formula_desc_cop.rb [hunk 20]
```diff
@@ -0,0 +1,84 @@
...
+            _receiver, call_name, args = *call_node
+            next if call_name != :desc || args.children[0].empty?
+            @formula_name = formula_name
+            @description = args.children[0]
+            @source_buffer = call_node.source_range.source_buffer
+            @line_number = call_node.loc.line
+            @line_begin_pos = call_node.source_range.source_buffer.line_range(call_node.loc.line).begin_pos
+            @desc_begin_pos = call_node.children[2].source_range.begin_pos
+            @call_node = call_node
```
Comment (MikeMcQuaid) [inline Library/Homebrew/rubocops/formula_desc_cop.rb]: I'm thinking some of this could be extracted into a method named `find_method_by_name(:desc)` which then returns the node or perhaps just the parameters? That's the type of thing I'm thinking with splitting this into smaller methods; have generic functions like `find_method_by_name` and later when needed `find_block_by_name` etc. which could take method `name` and optional `block_name` arguments and return e.g. the parameters or information as a hash. This should all live in `FormulaCop` so that it's not necessary to really understand RuboCop to implement a `FormulaCop` subclass and you can just call these documented, public methods. What do you think?

File: Library/.rubocop.yml (hunk 19 above)
Comment (reitermarkus) [inline Library/.rubocop.yml]: Also, as I said before, this will exclude every Formula in personal/custom Taps and Ruby files that are passed directly to `brew style`.

We need to exclude `HOMEBREW_LIBRARY_PATH` rather than include `Taps/**`.

Comment (MikeMcQuaid) [inline Library/.rubocop.yml]: I think just using `'**/Taps/**/*.rb'` may be sufficient? We don't want these applying to non-`HOMEBREW_LIBRARY_PATH` Ruby files.

Comment (reitermarkus) [inline Library/.rubocop.yml]: They won't, they will only match files with `class ***** < Formula` anyways.

File: Library/Homebrew/rubocops/formula_desc_cop.rb (hunk 20 above)
Comment (GauthamGoli) [inline Library/Homebrew/rubocops/formula_desc_cop.rb]: Yes, this makes sense. Wrapping rubocop specific code into documented methods would make it easier for other maintainers to add custom cops. But, I think, (as of now) its not possible to make writing custom cops to be completely rubocop agnostic as every possible requirement cannot be known beforehand, although we can try to cover most of the common cases,
And I think closures would be a good fit in this use case. I am pushing the code, please let me know further comments.
I apologize for the delay, as I was on a vacation.

File: Library/Homebrew/rubocops/extend/formula_cop.rb [hunk 21]
```diff
@@ -0,0 +1,67 @@
...
+  module Cop
+    module Homebrew
+      class FormulaCop < Cop
+        @registry = Cop.registry
+
+        def on_class(node)
+          # This method is called by RuboCop and is the main entry point
+          class_node, parent_class_node, body = *node
+          return unless a_formula_class?(parent_class_node)
+          audit_formula(node, class_node, parent_class_node, body)
```
Comment (reitermarkus) [inline Library/Homebrew/rubocops/extend/formula_cop.rb]: Add a `return unless respond_to?(:audit_formula)` here, so you can remove `Homebrew/FormulaCop` from the RuboCop config.

File: Library/Homebrew/rubocops/extend/formula_cop.rb [hunk 22]
```diff
@@ -0,0 +1,67 @@
+module RuboCop
+  module Cop
+    module Homebrew
+      class FormulaCop < Cop
+        @registry = Cop.registry
```
Comment (reitermarkus) [inline Library/Homebrew/rubocops/extend/formula_cop.rb]: What is this used for?

File: Library/Homebrew/rubocops/formula_desc_cop.rb [hunk 23]
```diff
@@ -0,0 +1,78 @@
...
+            # Check if formula's description size exceeds 80
+            desc_length = "#{formula_name}: #{description}".length
+            max_desc_length = 80
+            if desc_length > max_desc_length
+              column = desc_begin_pos - line_begin_pos
+              length = desc_size
+              offense_source_range = source_range(source_buffer, line_number, column, length)
+              desc_length_offense_msg = <<-EOS.undent
+              Description is too long. "name: desc" should be less than #{max_desc_length} characters.
+              Length is calculated as #{formula_name} + desc. (currently #{"#{formula_name}: #{description}".length})
```
Comment (reitermarkus) [inline Library/Homebrew/rubocops/formula_desc_cop.rb]: Indent these lines by 2 spaces.

File: Library/Homebrew/rubocops/extend/formula_cop.rb (hunk 22 above)
Comment (GauthamGoli) [inline Library/Homebrew/rubocops/extend/formula_cop.rb]: Ah! That's a hack. I hadn't found a solution for it yet. Without this line, there's an error ```undefined method `enlist' for nil:NilClass``` being thrown.

File: Library/Homebrew/rubocops/extend/formula_cop.rb (hunk 21 above)
Comment (GauthamGoli) [inline Library/Homebrew/rubocops/extend/formula_cop.rb]: Thanks! Added it.

File: Library/.rubocop.yml [hunk 24]
```diff
@@ -8,7 +8,15 @@ AllCops:
 
 require: ./Homebrew/rubocops.rb
 
+
 Homebrew/CorrectBottleBlock:
+  Include:
+    - '**/Taps/homebrew/**/*.rb'
```
Comment (MikeMcQuaid) [inline Library/.rubocop.yml]: Let's remove these includes and leave the logic in the cops themselves unless there's a way to do this that avoids duplicating this line for every cop.

File: Library/Homebrew/rubocops/extend/formula_cop.rb [hunk 25]
```diff
@@ -0,0 +1,68 @@
...
+
+        def on_class(node)
+          # This method is called by RuboCop and is the main entry point
+          class_node, parent_class_node, body = *node
+          return unless a_formula_class?(parent_class_node)
+          return unless respond_to?(:audit_formula)
+          audit_formula(node, class_node, parent_class_node, body)
+        end
+
+        def find_method(node, method_name)
```
Comment (MikeMcQuaid) [inline Library/Homebrew/rubocops/extend/formula_cop.rb]: `def find_node_method_by_name(node, method_name)` maybe?

File: Library/Homebrew/rubocops/bottle_block_cop.rb [hunk 26]
```diff
@@ -1,20 +1,31 @@
...
-          return if block_length(node).zero?
-          method, _args, body = *node
-          _keyword, method_name = *method
+      class CorrectBottleBlock < FormulaCop
+        MSG = "Use rebuild instead of revision in bottle block".freeze
 
-          return unless method_name == :bottle
-          check_revision?(body)
+        def audit_formula(_node, _class_node, _parent_class_node, formula_class_body_node)
+          check(formula_class_body_node)
```
Comment (MikeMcQuaid) [inline Library/Homebrew/rubocops/bottle_block_cop.rb]: Ideally this method would contain code like the following:
```ruby
def audit_formula
  bottle = find_block(:bottle)
  return unless bottle
  if method_called?(bottle, :revision)
    problem "Use rebuild instead of revision in a bottle block"
  end
end
```

The `find_block` and `method_called?` methods should be defined in `formula_cop.rb` and the various instance variables set by `on_class` or something it calls.

File: Library/Homebrew/rubocops/extend/formula_cop.rb [hunk 27]
```diff
@@ -0,0 +1,68 @@
...
+          # Executes a block for method named `method_name` called inside `node`
+          # Style violations are checked in the block and offenses can be reported
+          node.each_child_node(:send) do |call_node|
+            receiver, call_name, args = *call_node
+            next if call_name != method_name
+            yield receiver, call_name, args, call_node if block_given?
+          end
+        end
+
+        def node_line_begin_pos(node)
```
Comment (MikeMcQuaid) [inline Library/Homebrew/rubocops/extend/formula_cop.rb]: `def line_start_column(node)` maybe?

File: Library/Homebrew/rubocops/extend/formula_cop.rb [hunk 28]
```diff
@@ -0,0 +1,68 @@
...
+          # Returns the begin position of the node's line in source code
+          node.source_range.source_buffer.line_range(node.loc.line).begin_pos
+        end
+
+        def node_begin_pos(node)
```
Comment (MikeMcQuaid) [inline Library/Homebrew/rubocops/extend/formula_cop.rb]: `def start_column(node)` maybe?

File: Library/Homebrew/rubocops/extend/formula_cop.rb [hunk 29]
```diff
@@ -0,0 +1,68 @@
...
+          # Returns the begin position of the node in source code
+          node.source_range.begin_pos
+        end
+
+        def node_line_number(node)
```
Comment (MikeMcQuaid) [inline Library/Homebrew/rubocops/extend/formula_cop.rb]: `def line_number(node)` maybe?

File: Library/Homebrew/rubocops/extend/formula_cop.rb [hunk 30]
```diff
@@ -0,0 +1,68 @@
...
+          # Returns the line number of the node
+          node.loc.line
+        end
+
+        def class_node_name(node)
```
Comment (MikeMcQuaid) [inline Library/Homebrew/rubocops/extend/formula_cop.rb]: `def name(node)` maybe?

File: Library/Homebrew/rubocops/extend/formula_cop.rb [hunk 31]
```diff
@@ -0,0 +1,68 @@
...
+          # Returns the class node's name, nil if not a class node
+          node.const_name
+        end
+
+        def node_size(node)
```
Comment (MikeMcQuaid) [inline Library/Homebrew/rubocops/extend/formula_cop.rb]: `def size(node)` maybe?

File: Library/Homebrew/rubocops/extend/formula_cop.rb [hunk 32]
```diff
@@ -0,0 +1,68 @@
...
+          # Returns the node size in the source code
+          node.source_range.size
+        end
+
+        def get_source_buffer(node)
```
Comment (MikeMcQuaid) [inline Library/Homebrew/rubocops/extend/formula_cop.rb]: `source_buffer(node)` maybe?

File: Library/Homebrew/rubocops/extend/formula_cop.rb [hunk 33]
```diff
@@ -0,0 +1,68 @@
...
+          # Source buffer is required as an argument to report style violations
+          node.source_range.source_buffer
+        end
+
+        def get_str_content(node)
```
Comment (MikeMcQuaid) [inline Library/Homebrew/rubocops/extend/formula_cop.rb]: `string_content(node)` maybe?

File: Library/Homebrew/rubocops/extend/formula_cop.rb [hunk 34]
```diff
@@ -0,0 +1,68 @@
...
+          # Returns the string representation if node is of type str
+          node.str_content if node.type == :str
+        end
+
+        private
+
+        def a_formula_class?(parent_class_node)
```
Comment (MikeMcQuaid) [inline Library/Homebrew/rubocops/extend/formula_cop.rb]: `is_formula_class?(node)` maybe?

File: Library/Homebrew/rubocops/formula_desc_cop.rb [hunk 35]
```diff
@@ -0,0 +1,78 @@
...
+    module Homebrew
+      # This cop audits `desc` in Formulae
+      #
+      # - Checks for existence of `desc`
+      # - Checks if size of `desc` > 80
+      # - Checks if `desc` begins with an article
+      # - Checks for correct usage of `command-line` in `desc`
+      # - Checks if `desc` contains the formula name
+      class FormulaDesc < FormulaCop
+        def audit_formula(node, class_node, _parent_class_node, body)
```
Comment (MikeMcQuaid) [inline Library/Homebrew/rubocops/formula_desc_cop.rb]: Ideally this method would contain code like the following:
```ruby
def audit_formula
  desc_call = find_method_call(:desc)
  return unless desc_call
  desc = parameters(desc_call).first

  max_length = 80
  if desc.length > max_length
    problem <<-EOS.undent
      Description is too long. "name: desc" should be less than #{max_length} characters.
      Length is calculated as #{formula_name} + desc. (currently #{"#{formula_name}: #{desc}".length})
    EOS

  if desc =~ /(command ?line)/i
    problem "Description should use \"command-line\" instead of \"#{$1}\""
  end

 ... 

end
```

The `find_method_call` and `parameters?` methods should be defined in `formula_cop.rb` and the various instance variables set by `on_class` or something it calls and by e.g. `find_method_call` and `parameters` so `problem` can access e.g. the `source_range`.

Obviously this can (and will) vary but just giving you a heads up for the level of complexity I think would be ideal for these cops to hide the implementation details.

File: Library/Homebrew/rubocops/extend/formula_cop.rb (hunk 34 above)
Comment (reitermarkus) [inline Library/Homebrew/rubocops/extend/formula_cop.rb]: Without the `is_`, I'd say.

File: Library/Homebrew/rubocops/formula_desc_cop.rb (hunk 35 above)
Comment (GauthamGoli) [inline Library/Homebrew/rubocops/formula_desc_cop.rb]: `desc_call = find_method_call(:desc)` we are assuming that there would be only one method call. There might be multiple calls to the same method.  Should it return an array instead?

Comment (MikeMcQuaid) [inline Library/Homebrew/rubocops/formula_desc_cop.rb]: I think having `find_method_call` which always returns the `first` would be a sensible first step. If it's needed later we can have `find_method_calls` for e.g. checking use of things like `inreplace` and then loop over them.

Comment (GauthamGoli) [inline Library/Homebrew/rubocops/formula_desc_cop.rb]: While trying to implement the above I faced an issue. Just by having 
```ruby
if desc =~ /(command ?line)/i
  problem "Description should use \"command-line\" instead of \"#{$1}\""
end
```
its difficult to set all the instance variables required for pinpointing the `source_range` of the style violation. So I added a method `check_regex_match`. Now the code looks like 
```ruby
match_object = check_regex_match(desc, /(command ?line)/i)
if match_object
  problem "Description should use \"command-line\" instead of \"#{match_object}\""
end
```

Comment (MikeMcQuaid) [inline Library/Homebrew/rubocops/formula_desc_cop.rb]: How about:

```ruby
if match = regex_match_group(desc, /(command ?line)/i)
  problem "Description should use \"command-line\" instead of \"#{match_object}\""
end
```

File: Library/.rubocop.yml (hunk 24 above)
Comment (GauthamGoli) [inline Library/.rubocop.yml]: As @reitermarkus pointed out, due to `Include` in the `.rubocop.yml` the cops won't be run on personal/custom Taps and the files directly passed as arguments which do not match the pattern under `Include`.
 To resolve this issue I need some clarification.  `brew style` command, runs on all the files inside `HOMEBREW_LIBRARY_PATH` while `brew audit` runs only on the forumlae files. 
Do we want(or is it okay for) the custom cops to get executed under `brew audit` and `brew style` also?

Comment (MikeMcQuaid) [inline Library/.rubocop.yml]: > As @reitermarkus pointed out, due to Include in the .rubocop.yml the cops won't be run on personal/custom Taps and the files directly passed as arguments which do not match the pattern under Include.

I think it's OK for the taps to run on unnecessary files as they are still a pretty quick no-op.

> Do we want(or is it okay for) the custom cops to get executed under brew audit and brew style also?

Yeh, they should be executed by all of them I think (including `rubocop` obviously, just to be explicit)

Comment (GauthamGoli) [inline Library/.rubocop.yml]: There are classes in files
`compat/formula_specialties.rb`
`test/exceptions_spec.rb`
`test/support/fixtures/testball.rb`
`test/support/fixtures/testball_bottle.rb`
 inheriting from `Formula` in these files and are returning cop violations as they don't have formula `desc`, causing the build to fail. If `desc` can be added in the above files, then we don't have to repeat `Include` in `.rubocop.yml`

Comment (reitermarkus) [inline Library/.rubocop.yml]: > If desc can be added in the above files, then we don't have to repeat Include in .rubocop.yml

Why not just check if the file path contains `/Library/Homebrew/`, and exclude those? Didn't we have this approach already before?

Comment (GauthamGoli) [inline Library/.rubocop.yml]: Well, I was trying to keep the cop code and config code separate in their respective files. But turns out, keeping the `Include` logic in `FormulaCop` is more elegant. Thanks @reitermarkus and @MikeMcQuaid, now Personal Taps can be passed as arguments. 
I added the logic in Formula Cop and pushed the new code. I also used rebase to add another test in bottle_block_cop_spec in a past commit to pass coverage test.

**Comment by MikeMcQuaid** [comment]
This is looking pretty good to merge for me now. @reitermarkus, any thoughts?

**Comment by MikeMcQuaid** [comment]
Great work again @GauthamGoli! My suggestion for the next rules would be https://github.com/Homebrew/brew/blob/master/Library/Homebrew/dev-cmd/audit.rb#L251-L309; these are the ones I personally tend to wish to have in my editor the most.
